text/0715-construct-upgrade-support.md

```diff
@@ -0,0 +1,1251 @@
+
+```sh
+$ cdk construct-upgrade --unstable=construct-upgrade --target aws-cdk-lib.aws-dynamodb.TableV2 --custom-analysis demo-analysis.ts
+```
+
+An example of CDK built-in custom validation file content is as follows:
+
+```ts
+// demo-analysis.ts
+import { IRules, Rules } from '@aws-cdk/cli-custom-rules';
+
+class DemoDiffValidation implements DiffValidation {
+  execute(context: DiffContext) {
+    const v1Template = context.deployedTemplate;
+    const v2Template = context.newTemplate;
+    const changeSet = context.changeSet;
+
+    // Fail validation when `SkipReplicaDeletion` is not set on the deployed template
+    Object.entries(v1Template.Resources).forEach((item) => {
+      const [logicalId, resource] = item;
+      if (resource.Type === 'Custom::DynamoDBReplica') {
+        if (!resource.SkipReplicaRegion) {
+          // throw error
+        }
+      }
+    })
+
+    // Fail validation when `TableName` is not provided
+    Object.entries(v2Template.Resources).forEach((item) => {
+      const [logicalId, resource] = item;
+      if (resource.Type === 'AWS::DynamoDB::GlobalTable') {
+        if (!resource.TableName) {
+          // throw error
+        }
+      }
+    })
+  }
+}
+
+export default class DemoDiffAnalysis implements DiffAnalysis {
+  public readonly version = '1';
+
+  public init(host: IRules) {
+    host.registerCustomRules(new DemoDiffValidation());
+  }
+}
+```
+
+The construct upgrade CLI will load user-defined rules from `--custom-analysis` flag, triggering
+`cdk diff --custom-analysis demo-analysis.ts`, parse the error thrown as the `cdk diff` report,
+and fail the construct upgrade validations.
+
+Here is some future enhancements that we would make to the custom validations, see
+[Appendix E: Custom Validation Future Enhancement](#appendix-e-custom-validation-future-enhancements).
+
+## CI/CD Integration
+
+Enterprise developers often lack production AWS credentials, making pre-deployment
+validations (e.g., drift detection, change set analysis) impossible in production environments.
+
+At development time, here is how it works for common enterprise scenarios. Assuming developers have upgrade
+their CDK stack code to use new construct. They can use `dry-run` feature which will allow users to
+run the validations on the test/dev account without deploying.
+
+```sh
+$ cdk construct-upgrade --unstable=construct-upgrade --target aws-cdk-lib.aws-dynamodb.TableV2 --dry-run
+Performing construct upgrade validation...
+
+⚠️  Note: This is an experimental feature
+
+Validations:
+✅ No CloudFormation drift detected
+✅ Change set analysis complete
+```
+
+Depending on the deployment workflow and environment constraints, there are two common enterprise
+patterns:
+
+1. CI/CD-driven Deployment:
+For teams with access to CI/CD and production environments, developers commit the upgraded CDK stack
+code to version control. The CI/CD pipeline runs `cdk construct-upgrade` as part of the deployment process.
```

Comment:
Note that this explicitly won't work for deployment pipelines as used by Amazon, since they deploy CloudFormation stacks only and don't run arbitrary CLI commands.